File: Agentic Layer/architecture_contract.py
"""
Shared architecture JSON contract used across generation, cost, diagram, and IaC.
"""

from __future__ import annotations

import re
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator, model_validator

CloudProvider = Literal["aws", "azure", "gcp"]
_NODE_ID_RE = re.compile(r"^[A-Za-z][A-Za-z0-9_-]{0,127}$")


class ArchitectureContractError(ValueError):
    """Raised when architecture JSON does not match the shared contract."""


class ArchitectureNode(BaseModel):
    model_config = ConfigDict(extra="allow", str_strip_whitespace=True)

    id: str
    type: str
    label: str | None = None
    region: str | None = None
    attributes: dict[str, Any] = Field(default_factory=dict)

    @field_validator("id")
    @classmethod
    def node_id_valid(cls, value: str) -> str:
        if not _NODE_ID_RE.match(value):
            raise ValueError(
                "node id must start with a letter and contain only letters, digits, '_' or '-' (max 128 chars)"
            )
        return value

    @field_validator("type")
    @classmethod
    def node_type_non_empty(cls, value: str) -> str:
        if not value:
            raise ValueError("node type must not be empty")
        return value


class ArchitectureEdge(BaseModel):
    model_config = ConfigDict(extra="allow", populate_by_name=True, str_strip_whitespace=True)

    from_node: str = Field(alias="from")
    to_node: str = Field(alias="to")
    label: str | None = None

    @field_validator("from_node", "to_node")
    @classmethod
    def edge_endpoint_non_empty(cls, value: str) -> str:
        if not value:
            raise ValueError("edge endpoint must not be empty")
        return value


class ArchitectureDocument(BaseModel):
    model_config = ConfigDict(extra="allow", populate_by_name=True, str_strip_whitespace=True)

    title: str
    provider: CloudProvider | None = None
    schema_version: str = "1.0"
    nodes: list[ArchitectureNode]
    edges: list[ArchitectureEdge] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("title")
    @classmethod
    def title_non_empty(cls, value: str) -> str:
        if not value:
            raise ValueError("title must not be empty")
        return value

    @field_validator("schema_version")
    @classmethod
    def schema_version_non_empty(cls, value: str) -> str:
        if not value:
            raise ValueError("schema_version must not be empty")
        return value

    @field_validator("nodes")
    @classmethod
    def nodes_non_empty(cls, value: list[ArchitectureNode]) -> list[ArchitectureNode]:
        if not value:
            raise ValueError("nodes must contain at least one node")
        return value

    @model_validator(mode="after")
    def graph_integrity(self) -> "ArchitectureDocument":
        node_ids = [node.id for node in self.nodes]
        seen: set[str] = set()
        duplicates: set[str] = set()
        for node_id in node_ids:
            if node_id in seen:
                duplicates.add(node_id)
                continue
            seen.add(node_id)
        duplicate_list = sorted(duplicates)
        if duplicate_list:
            raise ValueError(f"duplicate node ids found: {', '.join(duplicate_list)}")

        valid_ids = set(node_ids)
        dangling: list[str] = []
        for edge in self.edges:
            if edge.from_node not in valid_ids or edge.to_node not in valid_ids:
                dangling.append(f"{edge.from_node}->{edge.to_node}")
        if dangling:
            raise ValueError(
                f"edges reference unknown nodes: {', '.join(dangling)}"
            )
        return self

    def to_wire_dict(self) -> dict[str, Any]:
        """Return normalized payload preserving edge aliases ('from'/'to')."""
        return self.model_dump(by_alias=True, exclude_none=True)


def parse_architecture_document(payload: Any) -> ArchitectureDocument:
    """Validate and parse raw payload to the shared architecture contract."""
    try:
        return ArchitectureDocument.model_validate(payload)
    except ValidationError as exc:
        details = "; ".join(
            f"{'.'.join(str(part) for part in err.get('loc', []))}: {err.get('msg')}"
            for err in exc.errors()
        )
        raise ArchitectureContractError(details or "Invalid architecture_json payload") from exc


def normalize_architecture_json(payload: Any) -> dict[str, Any]:
    """Validate and return normalized architecture JSON."""
    return parse_architecture_document(payload).to_wire_dict()

File: Connector/src/lib/architecture-contract.ts
export type ArchitectureProvider = 'aws' | 'azure' | 'gcp';

export interface ArchitectureNode {
  id: string;
  type: string;
  label?: string;
  region?: string;
  attributes: Record<string, unknown>;
  [key: string]: unknown;
}

export interface ArchitectureEdge {
  from: string;
  to: string;
  label?: string;
  [key: string]: unknown;
}

export interface ArchitectureJson {
  title: string;
  schema_version: string;
  provider?: ArchitectureProvider;
  nodes: ArchitectureNode[];
  edges: ArchitectureEdge[];
  metadata: Record<string, unknown>;
  [key: string]: unknown;
}

export interface ArchitectureValidationResult {
  valid: boolean;
  errors: string[];
  normalized?: ArchitectureJson;
}

const NODE_ID_RE = /^[A-Za-z][A-Za-z0-9_-]{0,127}$/;

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return null;
  return value as Record<string, unknown>;
}

function readOptionalString(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

export function validateArchitectureJson(input: unknown): ArchitectureValidationResult {
  const root = asRecord(input);
  if (!root) {
    return { valid: false, errors: ['architecture_json must be an object'] };
  }

  const errors: string[] = [];
  const title = readOptionalString(root.title);
  if (!title) errors.push('title is required and must be a non-empty string');

  const schemaVersion = readOptionalString(root.schema_version) || '1.0';
  const providerRaw = readOptionalString(root.provider);
  const provider = providerRaw ? providerRaw.toLowerCase() : undefined;
  if (provider && !['aws', 'azure', 'gcp'].includes(provider)) {
    errors.push('provider must be one of aws, azure, gcp when provided');
  }

  const rawNodes = Array.isArray(root.nodes) ? root.nodes : null;
  if (!rawNodes) {
    errors.push('nodes must be an array');
  } else if (rawNodes.length === 0) {
    errors.push('nodes must contain at least one node');
  }

  const nodes: ArchitectureNode[] = [];
  const nodeIds = new Set<string>();

  if (rawNodes) {
    rawNodes.forEach((item, idx) => {
      const node = asRecord(item);
      if (!node) {
        errors.push(`nodes[${idx}] must be an object`);
        return;
      }

      const id = readOptionalString(node.id);
      if (!id) {
        errors.push(`nodes[${idx}].id is required`);
        return;
      }
      if (!NODE_ID_RE.test(id)) {
        errors.push(`nodes[${idx}].id must match ${NODE_ID_RE.toString()}`);
      }
      if (nodeIds.has(id)) {
        errors.push(`duplicate node id found: ${id}`);
      }
      nodeIds.add(id);

      const type = readOptionalString(node.type);
      if (!type) {
        errors.push(`nodes[${idx}].type is required`);
      }

      const attributes = asRecord(node.attributes) || {};
      const normalizedNode: ArchitectureNode = {
        ...node,
        id,
        type: type || '',
        attributes,
      };
      const label = readOptionalString(node.label);
      if (label) normalizedNode.label = label;
      const region = readOptionalString(node.region);
      if (region) normalizedNode.region = region;
      nodes.push(normalizedNode);
    });
  }

  const rawEdges = Array.isArray(root.edges) ? root.edges : [];
  if (!Array.isArray(root.edges) && root.edges !== undefined) {
    errors.push('edges must be an array when provided');
  }

  const edges: ArchitectureEdge[] = [];
  rawEdges.forEach((item, idx) => {
    const edge = asRecord(item);
    if (!edge) {
      errors.push(`edges[${idx}] must be an object`);
      return;
    }
    const from = readOptionalString(edge.from);
    const to = readOptionalString(edge.to);
    if (!from) errors.push(`edges[${idx}].from is required`);
    if (!to) errors.push(`edges[${idx}].to is required`);
    if (!from || !to) return;

    if (!nodeIds.has(from) || !nodeIds.has(to)) {
      errors.push(`edges[${idx}] references unknown node(s): ${from} -> ${to}`);
    }

    const normalizedEdge: ArchitectureEdge = { ...edge, from, to };
    const label = readOptionalString(edge.label);
    if (label) normalizedEdge.label = label;
    edges.push(normalizedEdge);
  });

  const metadata = asRecord(root.metadata) || {};

  if (errors.length > 0) {
    return { valid: false, errors };
  }

  const normalized: ArchitectureJson = {
    ...root,
    title: title as string,
    schema_version: schemaVersion,
    nodes,
    edges,
    metadata,
  };
  if (provider === 'aws' || provider === 'azure' || provider === 'gcp') {
    normalized.provider = provider;
  }

  return {
    valid: true,
    errors: [],
    normalized,
  };
}
